drivers/pwm/pwm_nrfx.c

```diff
@@ -157,70 +149,82 @@ static int pwm_nrfx_set_cycles(const struct device *dev, uint32_t channel,
+		compare_value = PWM_NRFX_CH_COMPARE_MASK;
+	} else {
+		/* PWM generation needed. Check if the requested period matches
+		 * the one that is currently set, or the PWM peripheral can be
+		 * reconfigured accordingly.
+		 */
+		if (!pwm_period_check_and_set(dev, channel, period_cycles)) {
+			return -EINVAL;
 		}
+
+		compare_value = (uint16_t)(pulse_cycles >> data->prescaler);
+		needs_pwm = true;
 	}
 
-	data->current[channel] =
-		PWM_NRFX_CH_VALUE(pulse_cycles >> data->prescaler, inverted);
+	data->seq_values[channel] = PWM_NRFX_CH_VALUE(compare_value, inverted);
 
 	LOG_DBG("channel %u, pulse %u, period %u, prescaler: %u.",
 		channel, pulse_cycles, period_cycles, data->prescaler);
 
-	/* If this channel turns out to not need to be driven by the PWM
-	 * peripheral (it is off or fully on - duty 0% or 100%), set properly
+	/* If this channel does not need to be driven by the PWM peripheral
+	 * because its state is to be constant (duty 0% or 100%), set properly
 	 * the GPIO configuration for its output pin. This will provide
 	 * the correct output state for this channel when the PWM peripheral
-	 * is disabled after all channels appear to be inactive.
+	 * is stopped.
 	 */
-	if (!pwm_channel_is_active(channel, data)) {
+	if (!needs_pwm) {
 		uint32_t psel;
 
 		if (channel_psel_get(channel, &psel, config)) {
-			/* If pulse 0% and pin not inverted, set LOW.
-			 * If pulse 100% and pin inverted, set LOW.
-			 * If pulse 0% and pin inverted, set HIGH.
-			 * If pulse 100% and pin not inverted, set HIGH.
-			 */
-			bool pulse_0_and_not_inverted =
-				(pulse_cycles == 0U) && !inverted;
-			bool pulse_100_and_inverted =
-				(pulse_cycles == period_cycles) && inverted;
-			uint32_t value = (pulse_0_and_not_inverted ||
-					  pulse_100_and_inverted) ? 0 : 1;
-
-			nrf_gpio_pin_write(psel, value);
-		}
+			uint32_t out_level = (pulse_cycles == 0) ? 0 : 1;
 
-		if (!any_other_channel_is_active(channel, data)) {
-			nrfx_pwm_stop(&config->pwm, false);
+			if (inverted) {
+				out_level ^= 1;
+			}
+
+			nrf_gpio_pin_write(psel, out_level);
 		}
-	} else {
-		/* Since we are playing the sequence in a loop, the
-		 * sequence only has to be started when its not already
-		 * playing. The new channel values will be used
-		 * immediately when they are written into the seq array.
+	}
+
+	/* If the PWM generation is not needed for any channel (all are set
+	 * to constant inactive or active state), stop the PWM peripheral.
+	 * Otherwise, request a playback of the defined sequence so that
+	 * the PWM peripheral loads `seq_values` into its internal compare
+	 * registers and drives its outputs accordingly.
+	 */
+	if (!needs_pwm && !any_other_channel_needs_pwm(channel, data)) {
+		/* Don't wait here for the peripheral to actually stop. Instead,
+		 * ensure it is stopped before starting the next playback.
 		 */
-		if (was_stopped) {
-			/* Wait until PWM will be stopped and then start the
-			 * sequence.
+		nrfx_pwm_stop(&config->pwm, false);
```

Comment:
i wonder if we need to care about potential race condition here? What if there is a context switch here and another channel is set. In that case stop is initiated but `stop_requested` flag is not set. Second channel will not pend until pwm is stopped. What are the consequences?

Reply:
In consequence, a playback request made before the PWM actually stopped would be ignored.
But I don't think we should bother. The driver is not (and never was) thread-safe anyway.